manifest/dynamic_manifest.go

```diff
@@ -0,0 +1,226 @@
+	defer m.lk.RUnlock()
+	return m.manifest
+}
+
+func (m *DynamicManifestProvider) GpbftOptions() []gpbft.Option {
+	return m.manifest.GpbftOptions()
+}
+
+func (m *DynamicManifestProvider) SetManifestChangeCallback(mc OnManifestChange) {
+	m.onManifestChange = mc
+}
+
+// When a manifest configuration changes, a new network name
+// is set that depends on the manifest version of the previous version to avoid
+// overlapping previous configurations.
+func (m *DynamicManifestProvider) networkNameOnChange() gpbft.NetworkName {
+	return gpbft.NetworkName(string(m.manifest.NetworkName) + "/" + fmt.Sprintf("%d", m.manifest.Sequence))
+}
+
+func (m *DynamicManifestProvider) Run(ctx context.Context, errCh chan error) {
+	if m.onManifestChange == nil {
+		errCh <- xerrors.New("onManifestChange is nil. Callback for manifest change required")
+	}
+	go m.handleIncomingManifests(ctx, errCh)
+	m.handleApplyManifest(ctx, errCh)
+}
+
+func (m *DynamicManifestProvider) handleApplyManifest(ctx context.Context, errCh chan error) {
+	// add a timer for EC period
+	ticker := time.NewTicker(ManifestCheckTick)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			m.lk.Lock()
+			if m.nextManifest != nil {
+				ts, err := m.ec.GetHead(ctx)
+				if err != nil {
+					log.Errorf("error fetching chain head: %+v", err)
+					m.lk.Unlock()
+					continue
+				}
+
+				// if the upgrade epoch is reached or already passed.
+				if ts.Epoch() >= m.nextManifest.BootstrapEpoch {
+					log.Debugf("reached bootstrap epoch, triggering manifest change: %d", ts.Epoch())
+					// update the current manifest
+					prevManifest := m.manifest
+					m.manifest = *m.nextManifest
+					nn := m.networkNameOnChange()
+					m.manifest.NetworkName = nn
+					m.nextManifest = nil
+					// trigger manifest change callback.
+					go m.onManifestChange(ctx, prevManifest, errCh)
+					m.lk.Unlock()
+					continue
+				}
+			}
+			m.lk.Unlock()
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// listen to manifests being broadcast through the network.
+func (m *DynamicManifestProvider) handleIncomingManifests(ctx context.Context, errCh chan error) {
+	if err := m.setupManifestPubsub(); err != nil {
+		errCh <- xerrors.Errorf("setting up pubsub: %w", err)
+		return
+	}
+
+	manifestSub, err := m.manifestTopic.Subscribe()
+	if err != nil {
+		errCh <- xerrors.Errorf("subscribing to topic: %w", err)
+		return
+	}
+
+loop:
+	for {
+		select {
+		case <-ctx.Done():
+			break loop
+
+		default:
+			var msg *pubsub.Message
+			msg, err = manifestSub.Next(ctx)
+			if err != nil {
+				if ctx.Err() != nil {
+					break
+				}
+				log.Errorf("manifestPubsub subscription.Next() returned an error: %+v", err)
+				break
+			}
+			manifest, ok := msg.ValidatorData.(*Manifest)
+			if !ok {
+				log.Errorf("invalid manifestValidatorData: %+v", msg.ValidatorData)
+				continue
+			}
+
+			m.acceptNextManifest(manifest)
+		}
+	}
+
+	manifestSub.Cancel()
+	if err := m.teardownManifestPubsub(); err != nil {
```

Comment:
Cancel and tear down could go in a `differ` function to further simplify the code path if this function will keep doing its thing until context is done.